### PR TITLE
new ban validator functionality

### PR DESCRIPTION
**Pull request by Vitomir2**
- create a new contract and implement a ban function for the validators that will be banned;
- create a new internal function _withdraw to handle the withdraw process and re-use;
- write the tests for the new inspector contract;
- fix some other tests;

### CODE REVIEW

File: contracts/ValidatorSet/modules/Inspector/Inspector.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    function setReporterReward(uint256 newReward) public onlyOwner {
+        reporterReward = newReward;
+    }
+
+    // _______________ Private functions _______________
+
+    function _burnPenalty(address validator) private {
+        _burn(validator, validatorPenalty);
+        LiquidStaking._collectTokens(validator, validatorPenalty);
```
Comment (R-Santev) [inline contracts/ValidatorSet/modules/Inspector/Inspector.sol:57]: collect tokens must work differently. Will give you more information further.

File: test/ValidatorSet/Delegation.test.ts [hunk 2]
```diff
@@ -762,7 +762,7 @@ export function RunDelegationTests(): void {
 
         const maxTopUps = 52; // one cannot top-up more than 52 times
         for (let i = 0; i < maxTopUps; i++) {
-          const delegatingAmount = this.minDelegation.mul(i + 1).div(5);
+          const delegatingAmount = this.minDelegation.div(i + 1); //  mul(i + 1).div(5);
```
Comment (R-Santev) [inline test/ValidatorSet/Delegation.test.ts:765]: Why is this changed and a comment is added?

File: test/ValidatorSet/Inspector.test.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import { loadFixture, time } from "@nomicfoundation/hardhat-network-helpers";
+import * as hre from "hardhat";
+
+import { expect } from "chai";
+import { DENOMINATOR, VALIDATOR_STATUS, VESTING_DURATION_WEEKS, WEEK } from "../constants";
+import { registerValidator } from "../helper";
+
+export function RunInspectorTests(): void {
+  describe("Ban a validator", function () {
+    it("should revert if try to ban zero address", async function () {
```
Comment (R-Santev) [inline test/ValidatorSet/Inspector.test.ts:11]: Why banning the zero address is a problem in case it has some balance? I don't think we need this check in practise.

File: test/ValidatorSet/Inspector.test.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        .withArgs("UNREGISTERED_VALIDATOR");
+    });
+
+    it("should ban a validator successfully", async function () {
+      const { validatorSet } = await loadFixture(this.fixtures.stakedValidatorsStateFixture);
+
+      const penalizedValidator = this.signers.validators[0];
+      const stakedAmount = await validatorSet.balanceOf(penalizedValidator.address);
+
+      await validatorSet.connect(this.signers.governance).banValidator(this.signers.validators[0].address);
```
Comment (R-Santev) [inline test/ValidatorSet/Inspector.test.ts:33]: you can use the `penalizedValidator` variable here.

File: test/ValidatorSet/Inspector.test.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      const validatorData = await validatorSet.getValidator(penalizedValidator.address);
+
+      expect((await validatorSet.validators(penalizedValidator.address)).status, "status").to.be.equal(
+        VALIDATOR_STATUS.Banned
+      );
+      expect(validatorData.active, "active").to.be.equal(false);
+      expect(await validatorSet.balanceOf(penalizedValidator.address), "balanceOf").to.be.equal(0);
+      expect(await validatorSet.pendingWithdrawals(penalizedValidator.address), "withdrawals").to.be.equal(
+        stakedAmount.sub(validatorBanPenalty)
+      );
```
Comment (R-Santev) [inline test/ValidatorSet/Inspector.test.ts:45]: A check for some of the state changes that must occur is not made (e.g. stateSync)

File: test/ValidatorSet/Inspector.test.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    it("should ban a validator that has a vested staking position", async function () {
+      const { validatorSet, rewardPool } = await loadFixture(this.fixtures.stakedValidatorsStateFixture);
+
+      const newValidator = this.signers.accounts[9];
+
+      await registerValidator(validatorSet, this.signers.governance, newValidator);
+
+      const stakerValidatorSet = validatorSet.connect(newValidator);
+      await stakerValidatorSet.stakeWithVesting(VESTING_DURATION_WEEKS, {
```
Comment (R-Santev) [inline test/ValidatorSet/Inspector.test.ts:56]: Maybe you were able to use a fixture with a preset vesting position?

File: test/ValidatorSet/Inspector.test.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+      const vestingData = await rewardPool.positions(newValidator.address);
+      const nextTimestamp = vestingData.start.add(WEEK);
+      await time.setNextBlockTimestamp(nextTimestamp);
+
+      const validatorBanPenalty = await validatorSet.validatorPenalty();
+      const amountAfterBanPenalty = this.minStake.sub(validatorBanPenalty);
+      // hardcode the penalty percent by 0.3% a week
+      const bps = (VESTING_DURATION_WEEKS - 1) * 30;
+      const unstakePenalty = amountAfterBanPenalty.mul(bps).div(DENOMINATOR);
```
Comment (R-Santev) [inline test/ValidatorSet/Inspector.test.ts:68]: The process must be unstake -> unstake vested position penalty applied -> apply an additional penalty for the ban

File: test/ValidatorSet/Staking.test.ts [hunk 8]
```diff
@@ -72,7 +72,7 @@ export function RunStakingTests(): void {
     it("should not be able to unstake if there is insufficient staked balance", async function () {
       const { validatorSet } = await loadFixture(this.fixtures.stakedValidatorsStateFixture);
 
-      const unstakeInsufficientAmount = this.minStake.add(hre.ethers.utils.parseEther("5"));
+      const unstakeInsufficientAmount = this.minStake.mul(3);
```
Comment (R-Santev) [inline test/ValidatorSet/Staking.test.ts:75]: Why is this changed?

File: test/ValidatorSet/Staking.test.ts [hunk 9]
```diff
@@ -105,8 +105,8 @@ export function RunStakingTests(): void {
     it("should take pending unstakes into account", async function () {
       const { validatorSet } = await loadFixture(this.fixtures.stakedValidatorsStateFixture);
 
-      const unstakeInsufficientAmount = hre.ethers.utils.parseEther("2.5");
-      const stakeTooLowAmount = hre.ethers.utils.parseEther("1.5");
+      const unstakeInsufficientAmount = this.minStake.mul(3);
```
Comment (R-Santev) [inline test/ValidatorSet/Staking.test.ts:108]: Why is this changed?

File: test/helper.ts [hunk 10]
```diff
@@ -32,8 +32,8 @@ export async function initializeContext(context: any) {
   context.epochId = hre.ethers.BigNumber.from(1);
   context.epochSize = hre.ethers.BigNumber.from(64);
   context.epochReward = hre.ethers.utils.parseEther("0.0000001");
-  context.minStake = hre.ethers.utils.parseEther("1");
-  context.minDelegation = hre.ethers.utils.parseEther("1");
+  context.minStake = hre.ethers.utils.parseEther("1000");
```
Comment (R-Santev) [inline test/helper.ts:35]: Why is this changed?

**Review by R-Santev — CHANGES_REQUESTED** [review]

File: test/ValidatorSet/Delegation.test.ts (hunk 2 above)
Comment (Vitomir2) [reply test/ValidatorSet/Delegation.test.ts:765]: Forgot this, but will revert some tet changes.

File: test/ValidatorSet/Inspector.test.ts (hunk 3 above)
Comment (Vitomir2) [reply test/ValidatorSet/Inspector.test.ts:11]: Well, I just thought there is no way the zero address to become a validator, but yeah, even without this check, the functionality will work fine.

**Comment by Vitomir2** [comment]
@R-Santev some test files are modified because of tabulation and fixed small issues that caused tests to fail. Additionally, I have removed the `whitelisting` from the `registerValidator`, because it is a different purpose.

File: contracts/ValidatorSet/modules/Inspector/IInspector.sol [hunk 11]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+/**
+ * @notice Data type for the banned validators' withdrawals
+ * @param liquidTokens The amount of liquid tokens to be
```
Comment (R-Santev) [inline contracts/ValidatorSet/modules/Inspector/IInspector.sol:6]: description looks unfinished

File: contracts/ValidatorSet/modules/Inspector/Inspector.sol [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+     * @inheritdoc IInspector
+     */
+    function withdrawBannedFunds() public onlyBanned {
+        WithdrawalInfo memory withdrawalBalance = withdrawalBalances[msg.sender];
+
+        LiquidStaking._collectTokens(msg.sender, withdrawalBalance.liquidTokens);
+
+        _withdraw(msg.sender, withdrawalBalance.withdrawableAmount);
+
+        delete withdrawalBalances[msg.sender];
```
Comment (R-Santev) [inline contracts/ValidatorSet/modules/Inspector/Inspector.sol:93]: reentrancy attack vector

File: test/fixtures.ts [hunk 13]
```diff
@@ -239,7 +240,7 @@ async function vestingRewardsFixtureFunction(this: Mocha.Context) {
     this.epochSize
   );
 
-  await stakerValidatorSet.stake({ value: this.minStake });
+  await stakerValidatorSet.stake({ value: hre.ethers.utils.parseEther("1000") });
```
Comment (R-Santev) [inline test/fixtures.ts:243]: Why is this needed?

**Review by R-Santev — CHANGES_REQUESTED** [review]

File: contracts/ValidatorSet/modules/Inspector/Inspector.sol (hunk 12 above)
Comment (Vitomir2) [reply contracts/ValidatorSet/modules/Inspector/Inspector.sol:93]: Quite a bad error. Thanks for this one and I will make sure to take note!

File: test/fixtures.ts (hunk 13 above)
Comment (Vitomir2) [reply test/fixtures.ts:243]: Well, it is actually not needed as we discussed to just change the penalty and reporter reward in the Inspector contract.
